Many changes in the HangarAPI - v2.0.0

pavog opened the pull request:
A few days ago Hangar released some API changes.

From their [discord server](https://discord.gg/zvrAEbvJ4a):

> There are two important changes to the Hangar API: 
Firstly, all API routes previously containing author/slug should now be accessed just by its unique slug (e.g. api/v1/projects/{author}/{slug}/upload becomes api/v1/projects/{slug}/upload). This is to simplify getting and looking up projects by making project names unique. 
In the `projects/{slug}/upload API route, plugin dependencies should no longer include a namespace field for Hangar projects, but instead always use the name field for the now unique project name.
> 
> An actual break was done to the /projects endpoint, which had its searchWithRelevance parameter removed since it was not a really good metric as opposed to just sorting by stars, (recent) downloads, or (recent) views. By default, queries will now have the project with an exact name match as the first result, which you can toggle with the new prioritizeExactMatch parameter.
> 
> Otherwise, the old routes will continue to work where possible and simply redirect to the new ones, but you should still move away from the old as soon as you can. This should hopefully be the last large change to the API before a full release. :hatching_chick:


This results in the following changes:

### Changes in the generated code

**General**
 - Remove the author parameter and only use the slug in many APIs

**ChannelFlag**
 - Added `SENDS_NOTIFICATIONS`

**NamedPermission**
 - Removed `SEE_IP_ADDRESSES`

**PluginDependency**
 - Removed namespace property
 - Added platform property

**ProjectChannel**
 - Add description property

**ProjectsApi**
 - getProjects(): Replace parameter
                    `bool $order_with_relevance Whether projects should be sorted by the relevance to the given query (optional, default to true)`
                  with
                    `bool $prioritize_exact_match Whether to prioritize the project with an exact name match if present (optional, default to true)` See https://hangar.papermc.io/api-docs#/Projects/getProjects for more info

**ProjectSortingStrategy**
 - Removed `ONLY_RELEVANCE`

**UsersApi**
 - Add parameter `query` (search query) to methods `getAuthors()` and `getStaff()` to search for an user

**VersionsApi**
 - Remove author and only use slug
 - Deprecated old endpoints:
    - getVersions (/api/v1/projects/{author}/{slug}/versions)
    - getVersion (/api/v1/projects/{author}/{slug}/versions/{name})
    - downloadVersion -> renamed to downloadVersion1 (/api/v1/projects/{author}/{slug}/versions/{name}/{platform}/download)
    - getVersionStats (/api/v1/projects/{author}/{slug}/versions/{name}/stats)
    - getLatestReleaseVersion (/api/v1/projects/{author}/{slug}/latestrelease)
    - getLatestVersion (/api/v1/projects/{author}/{slug}/latest)
    - uploadVersion (/api/v1/projects/{author}/{slug}/upload)
 - New endpoints: 
    - uploadVersion (/api/v1/projects/{slug}/upload)
    - getLatestReleaseVersion (/api/v1/projects/{slug}/latestrelease)
    - getLatestVersion (/api/v1/projects/{slug}/latest)
    - getVersion (/api/v1/projects/{slug}/versions/{name})
    - getVersionStats (/api/v1/projects/{slug}/versions/{name}/stats)
    - getVersions (/api/v1/projects/{slug}/versions)
    - latestReleaseVersion
    - latestVersion


### Changes in our Client

**General**
 - Added the class `QueryableAndSortableUserList` that holds the properties `query` and `sort` and is used for staff and authors responses
 - Fixed all requests so that they do not use the author name any more and only use the slug

**HangerAPIClient**
 - getProject(): Removed parameter string $author
 - getProject(): Renamed parameter string $name to string $slug
 - getStaff(): Added parameter string $query, added parameter string|null $sort
 - getAuthors(): Added parameter string $query, added parameter string|null $sort
 - hasPermission(): Removed parameter ?string $owner
 - hasPermissions(): Removed parameter ?string $owner
 - getDailyProjectStats(): Removed parameter string $owner
 - editProjectPage(): Removed parameter string $author
 - getProjectMainPage(): Removed parameter string $author
 - getProjectPage(): Removed parameter string $author
 - editProjectMainPage(): Removed parameter string $author

**ProjectPage**
 - Removed property `protected string $owner`



I bumped the version to 2.0.0 because this PR contains some changes (like removing and adding some properties), although Hangar did not change its API version.

pavog committed:
Update generated code

pavog committed:
Update client to match generated code + introduce QueryableAndSortableUserList and use it for StaffList and AuthorList

pavog committed:
Update tests

pavog committed:
Bump version to 2.0.0

JulianVennen commented on lib/Client/HangarAPIClient.php:
```diff
@@ -492,11 +477,13 @@ public function getProjectsPinnedByUser(string $username): array
      * @throws ApiException
      */
-    public function getStaff(?RequestPagination $pagination = null): StaffList
+    public function getStaff(string $query, ?RequestPagination $pagination = null, ?string $sort = null): StaffList
```
You could probably default initialize query as "".
It would also be better to have an enum for the sort parameter

pavog committed:
Use empty string as default value for search query in getAuthors and getStaff

pavog commented:
Migration guide from 1.0.0 to 2.0.0
===============================

### `HangarAPIClient->getStaff()` method signature changed (added parameter `$query` and added parameter `$sort`)
from `public function getStaff(?RequestPagination $pagination = null): StaffList`
to `public function getStaff(string $query = "", ?RequestPagination $pagination = null, ?string $sort = null): StaffList`
- If you don't use the argument `$pagination` or pass it as a named argument, you must not change anything
- Please update your function calls to include the argument `$query` if do not want to use the default value ("" = all) or if you use the argument `$pagination` and do not pass it as a named argument

### `HangarAPIClient->getAuthors()` method signature changed (added parameter `$query` and added parameter `$sort`)
from `public function getAuthors(?RequestPagination $pagination = null): AuthorList`
to `public function getAuthors(string $query = "", ?RequestPagination $pagination = null, ?string $sort = null): AuthorList`
- If you don't use the argument `$pagination` or pass it as a named argument, you must not change anything
- Please update your function calls to include the argument `$query` if do not want to use the default value ("" = all) or if you use the argument `$pagination` and do not pass it as a named argument

### `HangarAPIClient->getProject()` method signature changed (removed parameter `$author`)
from `public function getProject(string $author, string $name): Project`
to `public function getProject(string $slug): Project`
- Please remove the argument `$author` from your function calls

### `HangarAPIClient->getProjectMainPage()` method signature changed (removed parameter `$author`)
from `public function getProjectMainPage(string $author, string $slug): ProjectPage`
to `public function getProjectMainPage(string $slug): ProjectPage`
- Please remove the argument `$author` from your function calls

### `HangarAPIClient->getProjectPage()` method signature changed (removed parameter `$author`)
from `public function getProjectPage(string $author, string $slug, string $path): ProjectPage`
to `public function getProjectPage(string $slug, string $path): ProjectPage`
- Please remove the argument `$author` from your function calls

### `HangarAPIClient->editProjectMainPage()` method signature changed (removed parameter `$author`)
from `public function editProjectMainPage(string $author, string $slug, string $content): ProjectPage`
to `public function editProjectMainPage(string $slug, string $content): ProjectPage`
- Please remove the argument `$author` from your function calls

### `HangarAPIClient->editProjectPage()` method signature changed (removed parameter `$author`)
from `public function editProjectPage(string $author, string $slug, string $path, string $content): ProjectPage`
to `public function editProjectPage(string $slug, string $path, string $content): ProjectPage`
- Please remove the argument `$author` from your function calls

### `HangarAPIClient->hasPermissions()` method signature changed (removed parameter `$owner`)
from `public function hasPermissions(array $permissions, ?string $owner = null, ?string $project = null): bool`
to `public function hasPermissions(array $permissions, ?string $project = null): bool`
- Please remove the argument `$owner` from your function calls

### `HangarAPIClient->hasPermission()` method signature changed (removed parameter `$owner`)
from `public function hasPermission(string $permission, ?string $owner = null, ?string $project = null): bool`
to `public function hasPermission(string $permission, ?string $project = null): bool`
- Please remove the argument `$owner` from your function calls

### `HangarAPIClient->getDailyProjectStats()` method signature changed (removed parameter `$owner`)
from `public function getDailyProjectStats(string $owner, string $slug, DateTime $from, ?DateTime $to = null): array`
to `public function getDailyProjectStats(string $slug, DateTime $from, ?DateTime $to = null): array`
- Please remove the argument `$owner` from your function calls
